File: go/iriq/extractor.go
package iriq

import (
	"regexp"
	"strings"
	"unicode/utf8"
)

// Extractor pulls IRIs out of free text. Scheme-anchored — only URLs whose
// scheme appears explicitly are extracted; scheme-less hosts like "foo.com/x"
// are extracted only when SchemeLess is true (default).
type Extractor struct {
	SchemeLess bool
}

func NewExtractor() *Extractor { return &Extractor{SchemeLess: true} }

var (
	extractorSchemes        = []string{"https", "http", "ftp", "wss", "ws"}
	extractorSchemelessTLDs = []string{"com", "org", "net", "io", "ai", "dev", "co", "app", "gov", "edu"}

	// CJK/Unicode brackets and curly quotes that almost always terminate a URL
	// in source text. Mirrors NON_ASCII_BOUNDARY in the Ruby implementation.
	extractorNonASCIIBoundary = uniqueRunes(
		"」』）】〉》〕〗〙〛｠｝］＞" +
			"「『（【〈《〔〖〘〚｟｛［＜" +
			"“”‘’„‟‚«»‹›",
	)
)

// extractorURLChars is the character class allowed inside an extracted URL:
// everything except ASCII whitespace, angle brackets, quotes, backtick,
// comma, and the listed Unicode boundary chars.
var extractorURLChars = `[^\s<>"'` + "`" + `,` + regexpCharClassEscape(extractorNonASCIIBoundary) + `]+`

var (
	// Scheme-anchored only (used when SchemeLess is false).
	candidateRE = regexp.MustCompile(
		`(?:` +
			`(?i:` + strings.Join(extractorSchemes, "|") + `)://` + extractorURLChars +
			`|` +
			`urn:[a-zA-Z0-9][a-zA-Z0-9\-]{0,30}:` + extractorURLChars +
			`)`,
	)

	// Combined pattern: scheme-anchored OR urn OR scheme-less.
	combinedRE = regexp.MustCompile(
		`(?:` +
			`(?i:` + strings.Join(extractorSchemes, "|") + `)://` + extractorURLChars +
			`|` +
			`urn:[a-zA-Z0-9][a-zA-Z0-9\-]{0,30}:` + extractorURLChars +
			`|` +
			`(?:[a-zA-Z0-9](?:[a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+(?i:` +
			strings.Join(extractorSchemelessTLDs, "|") + `)/` + extractorURLChars +
			`)`,
	)

	trailingPunctRE = regexp.MustCompile(`[.,;:!?'"\x{2018}\x{2019}\x{201C}\x{201D}]+$`)
)

var bracketPairs = map[rune]rune{')': '(', ']': '[', '}': '{'}

// Extract scans text and returns parsed IRIs in source order. Unparseable
// candidates are silently dropped.
func (e *Extractor) Extract(text string) []*Identifier {
	if text == "" {
		return nil
	}
	pattern := candidateRE
	if e.SchemeLess {
		pattern = combinedRE
	}
	out := make([]*Identifier, 0, 4)
	for _, loc := range pattern.FindAllStringIndex(text, -1) {
		if !leftBoundaryOK(text, loc[0], e.SchemeLess) {
			continue
		}
		candidate := text[loc[0]:loc[1]]
		trimmed := trimCandidate(candidate)
		if trimmed == "" {
			continue
		}
		iri, err := Parse(trimmed)
		if err != nil {
			continue
		}
		out = append(out, iri)
	}
	return out
}

// ExtractStrings is like Extract but returns canonical strings, deduplicated,
// preserving first-seen order.
func (e *Extractor) ExtractStrings(text string) []string {
	seen := map[string]struct{}{}
	out := make([]string, 0)
	for _, iri := range e.Extract(text) {
		c := iri.Canonical()
		if _, ok := seen[c]; ok {
			continue
		}
		seen[c] = struct{}{}
		out = append(out, c)
	}
	return out
}

// leftBoundaryOK enforces what Ruby's negative lookbehind (?<![\w/.@]) or
// (?<![\w/]) does. RE2 has no lookbehind, so we re-check after the match.
//
// Scheme-only mode uses [\w/]; combined mode also excludes . and @ to prevent
// extracting from inside emails and longer hostnames like "a.foo.com".
func leftBoundaryOK(text string, start int, schemeless bool) bool {
	if start == 0 {
		return true
	}
	r, _ := utf8.DecodeLastRuneInString(text[:start])
	if r == utf8.RuneError {
		return true
	}
	if isWordRune(r) {
		return false
	}
	if r == '/' {
		return false
	}
	if schemeless && (r == '.' || r == '@') {
		return false
	}
	return true
}

// isWordRune mirrors Ruby's Unicode-aware \w in /xu mode: letters, marks,
// numbers, and underscore.
func isWordRune(r rune) bool {
	if r == '_' {
		return true
	}
	return unicodeIsLetterOrDigit(r)
}

func trimCandidate(candidate string) string {
	s := candidate
	for {
		before := s
		s = trailingPunctRE.ReplaceAllString(s, "")
		for close, open := range bracketPairs {
			for len(s) > 0 && lastRune(s) == close && runeCount(s, close) > runeCount(s, open) {
				s = trimLastRune(s)
			}
		}
		if s == before {
			return s
		}
	}
}

func uniqueRunes(s string) string {
	seen := map[rune]struct{}{}
	var out []rune
	for _, r := range s {
		if _, ok := seen[r]; ok {
			continue
		}
		seen[r] = struct{}{}
		out = append(out, r)
	}
	return string(out)
}

// regexpCharClassEscape escapes the small set of characters that have special
// meaning inside an RE2 character class: ], \, ^, -.
func regexpCharClassEscape(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch r {
		case ']', '\\', '^', '-':
			b.WriteByte('\\')
		}
		b.WriteRune(r)
	}
	return b.String()
}

File: lib/iriq/extractor.rb
module Iriq
  # Pulls IRIs out of free text. Scheme-anchored — only URLs whose scheme
  # appears explicitly are extracted (scheme-less hosts like "foo.com/x" are
  # too noisy to disambiguate from prose).
  #
  #   Iriq::Extractor.new.extract("Visit https://foo.com today.")
  #   # => [#<Iriq::Identifier https://foo.com>]
  #
  # Design draws on twitter-text and GFM autolink rules: scheme anchoring,
  # iterative trailing-punct trim, balanced-paren preservation.
  class Extractor
    SCHEMES = %w[https http ftp wss ws].freeze

    # Conservative TLD allow-list for scheme-less extraction. Limited to a
    # small set of very common TLDs to keep false-positive rate low. A
    # scheme-less candidate ALSO requires a `/path` to count, so plain
    # `foo.com` in prose still won't match — only `foo.com/something`.
    SCHEMELESS_TLDS = %w[com org net io ai dev co app gov edu].freeze

    # Boundary chars — a URL ends at any of these (whitespace, angle
    # brackets, quotes, backtick).
    BOUNDARY = %r{[\s<>"'`]}.freeze

    # Non-ASCII Unicode brackets and quotation marks that almost always
    # terminate a URL in source text (e.g. `「URL」`). ASCII brackets are NOT
    # listed here — those stay inside the URL match so the balanced-paren
    # trim step can handle them (Wikipedia URLs like /Foo_(bar) survive).
    NON_ASCII_BOUNDARY = (
      "」』）】〉》〕〗〙〛｠｝］＞" +  # CJK closing brackets
      "「『（【〈《〔〖〘〚｟｛［＜" +  # CJK opening brackets
      "“”‘’„‟‚«»‹›"                 # Unicode quotation marks
    ).chars.uniq.join.freeze

    URL_CHAR_CLASS = %{[^\\s<>"'`,#{NON_ASCII_BOUNDARY}]+}.freeze

    CANDIDATE_RE = %r{
      (?<![\w/])                                                    # not mid-word, not mid-path
      (?:
        (?i:#{SCHEMES.join("|")})://#{URL_CHAR_CLASS}               # absolute URL
        |
        urn:[a-zA-Z0-9][a-zA-Z0-9\-]{0,30}:#{URL_CHAR_CLASS}        # urn:NID:NSS
      )
    }xu.freeze

    # Scheme-less pattern: opt-in, conservative. Requires a host whose TLD
    # is in SCHEMELESS_TLDS AND a `/path` to disambiguate from prose. The
    # host part allows ASCII labels separated by dots; no Unicode hosts
    # (those are too easily confused with prose).
    SCHEMELESS_RE = %r{
      (?<![\w/.@])                                                   # boundary; not after @ (avoid emails)
      (?:[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+            # label(s)
      (?i:#{SCHEMELESS_TLDS.join("|")})                              # TLD
      /#{URL_CHAR_CLASS}                                             # required path
    }xu.freeze

    # Punctuation that's almost always sentence punctuation rather than part
    # of a URL when it appears at the trailing edge.
    TRAILING_PUNCT_RE = /[.,;:!?'"‘’“”]+\z/u.freeze

    # Unmatched closing brackets that should be trimmed.
    BRACKET_PAIRS = { ")" => "(", "]" => "[", "}" => "{" }.freeze

    def initialize(scheme_less: true)
      @scheme_less = scheme_less
    end

    def extract(text)
      return [] if text.nil? || text.empty?

      candidates = scan_candidates(text)
      candidates.filter_map do |candidate|
        trimmed = trim(candidate)
        next nil if trimmed.empty?

        begin
          Parser.parse(trimmed)
        rescue ParseError
          nil
        end
      end
    end

    # Same as extract but returns only canonical strings, deduplicated,
    # preserving first-seen order.
    def extract_strings(text)
      seen = {}
      extract(text).each { |iri| seen[iri.canonical] ||= true }
      seen.keys
    end

    private

    # Combine scheme + (optionally) scheme-less matches in source order.
    def scan_candidates(text)
      matches = []
      text.scan(CANDIDATE_RE) { matches << [Regexp.last_match.begin(0), Regexp.last_match[0]] }
      if @scheme_less
        text.scan(SCHEMELESS_RE) { matches << [Regexp.last_match.begin(0), Regexp.last_match[0]] }
      end
      matches.sort_by(&:first).map(&:last)
    end

    # Iteratively strip sentence punctuation and unmatched closing brackets
    # until the candidate stabilizes.
    def trim(candidate)
      s = candidate.dup
      loop do
        before = s
        s = s.sub(TRAILING_PUNCT_RE, "")
        BRACKET_PAIRS.each do |close, open|
          while s.end_with?(close) && s.count(close) > s.count(open)
            s = s[0...-1]
          end
        end
        break if s == before
      end
      s
    end
  end
end
